demos/source/aws_iot_demo_jobs.c

```diff
@@ -0,0 +1,533 @@
+ */
+int32_t RunJobsDemo( bool awsIotMqttMode,
+                     const char * pIdentifier,
+                     void * pNetworKServerInfo,
+                     void * pNetworkCredentialInfo,
+                     const IotNetworkInterface_t * pNetworkInterface );
+
+/**
+ * @brief Initialize The Jobs Demo.
+ *
+ * Initialize MQTT module then the Jobs module. If the Jobs module fails
+ * to initialize, MQTT cleanup is invoked.
+ *
+ * @return EXIT_SUCCESS if initialization succeeds. Otherwise EXIT_FAILURE.
+ */
+static int32_t _initDemo( void )
+{
+    int32_t status = EXIT_SUCCESS;
+    IotMqttError_t mqttInitStatus = IOT_MQTT_SUCCESS;
+    AwsIotJobsError_t jobsInitStatus = AWS_IOT_JOBS_SUCCESS;
+
+    mqttInitStatus = IotMqtt_Init();
+
+    if( mqttInitStatus != IOT_MQTT_SUCCESS )
+    {
+        status = EXIT_FAILURE;
+    }
+
+    if( status == EXIT_SUCCESS )
+    {
+        jobsInitStatus = AwsIotJobs_Init( AWS_IOT_DEMO_JOBS_MQTT_TIMEOUT );
+
+        if( jobsInitStatus != AWS_IOT_JOBS_SUCCESS )
+        {
+            IotMqtt_Cleanup();
+            status = EXIT_FAILURE;
+        }
+    }
+
+    return status;
+}
+
+/**
+ * @brief The callback when a new job is posted. It signals main via a semaphore.
+ */
+static void _jobsCallback( void * param1,
+                           AwsIotJobsCallbackParam_t * pCallbackInfo )
+{
+    bool success = false;
+    const char * pJsonValue = NULL;
+    size_t jsonValueLength = 0;
+    IotSemaphore_t * pWaitSem = ( IotSemaphore_t * ) param1;
+
+    IotLogInfo( "Job Doc Len : %d", pCallbackInfo->u.callback.documentLength );
+    IotLogInfo( "Job Document: %s", pCallbackInfo->u.callback.pDocument );
+
+    /* Get the Job ID */
+    success =
+        IotJsonUtils_FindJsonValue( pCallbackInfo->u.callback.pDocument,
+                                    pCallbackInfo->u.callback.documentLength,
+                                    JOB_ID_KEY,
+                                    JOB_ID_KEY_LENGTH,
+                                    &pJsonValue,
+                                    &jsonValueLength );
+    if( success )
+    {
+        IotLogInfo( "New Job: %.*s", jsonValueLength, pJsonValue );
+        memcpy( _pJobId, pJsonValue, jsonValueLength );
+        _JobIdLength = jsonValueLength;
+    }
+    else
+    {
+        IotLogError( "Failed to parse JobID from Notify Next." );
+    }
+
+    /* Get the Job Document */
+    success =
+        IotJsonUtils_FindJsonValue( pCallbackInfo->u.callback.pDocument,
+                                    pCallbackInfo->u.callback.documentLength,
+                                    JOB_DOC_KEY,
+                                    JOB_DOC_KEY_LENGTH,
+                                    &pJsonValue,
+                                    &jsonValueLength );
+
+    if( success )
+    {
+        IotLogInfo( "New Job Doc: %.*s", jsonValueLength, pJsonValue );
+        memcpy( _jobDoc, pJsonValue, jsonValueLength );
```

Comment:
If jsonValueLength is longer than _jobDoc's length, then _jobDoc will contain a partial and probably invalid document.  This needs to be handled somehow.